assets/js/components/DashboardNavigation.js

```diff
@@ -99,76 +99,128 @@ export default function DashboardNavigation() {
 	);
 
+	const handleSelect = ( chipID ) => {
+		const hash = chipID;
+		if ( hash ) {
+			global.scrollTo( {
+				top:
+					hash !== ANCHOR_ID_TRAFFIC
+						? getContextScrollTop( `#${ hash }`, breakpoint )
+						: 0,
+				behavior: 'smooth',
+			} );
+		} else {
+			global.history.replaceState(
+				{},
+				'',
+				removeQueryArgs( global.location.href )
+			);
+			setSelectedID( hash );
+		}
+	};
+
 	useMount( () => {
-		if ( global.location.hash !== '' ) {
-			setTimeout( () => {
-				const hash = global.location.hash.substr( 1 );
-				global.scrollTo( {
-					top:
-						hash !== ANCHOR_ID_TRAFFIC
-							? getContextScrollTop( `#${ hash }`, breakpoint )
-							: 0,
-					behavior: 'smooth',
-				} );
-			}, 10 );
+		const { hash } = global.location;
+		if ( ! hash ) {
+			return;
 		}
+
+		setTimeout( () => {
+			global.scrollTo( {
+				top:
+					hash.substring( 1 ) !== ANCHOR_ID_TRAFFIC
+						? getContextScrollTop( hash, breakpoint )
+						: 0,
+				behavior: 'smooth',
+			} );
+		}, 25 );
 	} );
 
+	useEffect( () => {
+		const onScroll = () => {
+			const entityHeader = document
+				.querySelector( '.googlesitekit-entity-header' )
+				?.getBoundingClientRect()?.bottom;
+			const navigation = document
+				.querySelector( '.googlesitekit-navigation' )
+				?.getBoundingClientRect()?.bottom;
+			const margin = 20;
+
+			const areas = [
+				...( showTraffic ? [ ANCHOR_ID_TRAFFIC ] : [] ),
+				...( showContent ? [ ANCHOR_ID_CONTENT ] : [] ),
+				...( showSpeed ? [ ANCHOR_ID_SPEED ] : [] ),
+				...( showMonetization ? [ ANCHOR_ID_MONETIZATION ] : [] ),
+			];
+
+			let closest;
+			let closestID = ANCHOR_ID_TRAFFIC;
+
+			for ( const areaID of areas ) {
+				const area = document.getElementById( areaID );
+				if ( ! area ) {
+					continue;
+				}
+
+				const top =
+					area.getBoundingClientRect().top -
+					margin -
+					( entityHeader || navigation || 0 );
+
+				if ( top < 0 && ( closest === undefined || closest < top ) ) {
+					closest = top;
+					closestID = areaID;
+				}
+			}
+
+			const { hash } = global.location;
+			if ( closestID !== hash?.substring( 1 ) ) {
+				global.history.replaceState( {}, '', `#${ closestID }` );
+				setSelectedID( closestID );
+			}
+		};
+
+		const throttledOnScroll = throttle( onScroll, 50 );
+		global.addEventListener( 'scroll', throttledOnScroll );
+
+		throttledOnScroll();
+
+		return () => {
+			global.removeEventListener( 'scroll', throttledOnScroll );
+		};
+	}, [ showTraffic, showContent, showSpeed, showMonetization ] );
+
 	return (
-		<ChipSet
-			className="googlesitekit-navigation"
-			selectedChipIds={ selectedIds }
-			handleSelect={ handleSelect }
-			choice
-		>
+		<div className="googlesitekit-navigation mdc-chip-set">
 			{ showTraffic && (
 				<Chip
 					id={ ANCHOR_ID_TRAFFIC }
 					label={ __( 'Traffic', 'google-site-kit' ) }
 					leadingIcon={ <NavTrafficIcon width="18" height="16" /> }
+					onClick={ handleSelect.bind( null, ANCHOR_ID_TRAFFIC ) }
```

Comment:
We should use `useCallback` for these functions as is our convention.

Reply:
@hussain-t `handleSelect` now uses `useCallback` but using `bind` here will create a new function on every render. Let's extract an additional callback for each (e.g. `handleSelectTraffic`, `handleSelectContent`) here.